cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -321,170 +338,202 @@ private void recordProgramStart(ProgramId programId, String pid, long startTs, S
-      addWorkflowNodeState(programId, pid, systemArgs, ProgramRunStatus.RUNNING, null);
+      addWorkflowNodeState(programId, pid, systemArgs, ProgramRunStatus.RUNNING, null, sourceId);
     }
 
     // Delete the old run record
+    MDSKey key = getProgramKeyBuilder(STATUS_TYPE_MAP.get(existing.getStatus()), programId).add(pid).build();
     deleteAll(key);
 
-    // Build the key for TYPE_RUN_RECORD_STARTING
+    // Build the key for TYPE_RUN_RECORD_STARTED
     key = keyBuilder.add(pid).build();
     // The existing record's properties already contains the workflowRunId
     RunRecordMeta meta = new RunRecordMeta(pid, existing.getStartTs(), runTs, null, ProgramRunStatus.RUNNING,
-                                           existing.getProperties(), systemArgs, twillRunId);
+                                           existing.getProperties(), systemArgs, twillRunId, sourceId);
     write(key, meta);
   }
 
-  public void recordProgramSuspend(ProgramId program, String pid) {
-    recordProgramSuspendResume(program, pid, "suspend");
-  }
-
-  public void recordProgramResumed(ProgramId program, String pid) {
-    recordProgramSuspendResume(program, pid, "resume");
+  public void recordProgramSuspend(ProgramId programId, String pid, byte[] sourceId) {
+    List<RunRecordMeta> existingRecords = getRuns(programId, pid);
+    boolean isValid = validateExistingRecords(existingRecords, programId, pid, sourceId, "suspend",
+                                              ImmutableSet.of(ProgramRunStatus.STARTING, ProgramRunStatus.RUNNING),
+                                              ImmutableSet.of(ProgramRunStatus.SUSPENDED));
+    if (!isValid) {
+      // Skip recording start if the existing records are not valid
+      return;
+    }
+    // Only one record in the valid existing records
+    RunRecordMeta existing = existingRecords.get(0);
+    recordProgramSuspendResume(programId, pid, sourceId, existing, "suspend");
+  }
+
+  public void recordProgramResumed(ProgramId programId, String pid, byte[] sourceId) {
+    List<RunRecordMeta> existingRecords = getRuns(programId, pid);
+    boolean isValid = validateExistingRecords(existingRecords, programId, pid, sourceId, "suspend",
+                                              ImmutableSet.of(ProgramRunStatus.SUSPENDED),
+                                              ImmutableSet.of(ProgramRunStatus.STARTING, ProgramRunStatus.RUNNING));
+    if (!isValid && !existingRecords.isEmpty()) {
+      // Skip recording start if the existing records are not valid
+      return;
+    }
+    RunRecordMeta existing = null;
+    // Only if existingRecords is empty & upgrade is not complete & the version is default version,
+    // also try to get the record without the version string
+    if (existingRecords.isEmpty()) {
+      if (!upgradeComplete.get() && programId.getVersion().equals(ApplicationId.DEFAULT_VERSION)) {
+        MDSKey key = getVersionLessProgramKeyBuilder(TYPE_RUN_RECORD_SUSPENDED, programId)
```

Comment:
Copied this from existing code. However, we actually only upgrade record of type `TYPE_RUN_RECORD_COMPLETED` with the assumption all programs are stopped before upgrade. However, here the assumption is that there are still records of type `TYPE_RUN_RECORD_SUSPENDED`. We should only have one assumption. Still asking Gokul for confirmation